File: src/main/node/net/x3djsonld/data/JoeExample.js
var java = require('java');
java.asyncOptions = {
  asyncSuffix: undefined,     // Don't generate node-style methods taking callbacks
  syncSuffix: "",              // Sync methods use the base name(!!)
  promiseSuffix: "Promise",   // Generate methods returning promises, using the suffix Promise.
  promisify: require('util').promisify, // Needs Node.js version 8 or greater, see comment below
  ifReadOnlySuffix: "_alt"
};
var autoclass = require('./X3Dautoclass');

// Javadoc annotations follow, see below for source.
/**
 * <p> Example scene JoeExample. </p>
 <p> Related links: JoeExample.java source, <a href="https://www.web3d.org/x3d/content/examples/X3dResources.html" target="_blank">X3D Resources</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dSceneAuthoringHints.html" target="_blank">X3D Scene Authoring Hints</a> and <a href="https://www.web3d.org/x3d/content/X3dTooltips.html" target="_blank">X3D Tooltips</a>. </p>
	<table style="color:black; border:0px solid; border-spacing:10px 0px;" summary="Scene Metadata">
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center; padding:10px 0px;"><i>meta tags</i></td>
			<td style="text-align:left;   padding:10px 0px;">&nbsp; Document Metadata </td>
		</tr>

		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> filename </i> </td>
			<td> Scripting.x3d </td>
		</tr>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center;" colspan="2">  &nbsp; </td>
		</tr>
	</table>

	<p>
		This program uses the
		<a href="https://www.web3d.org/specifications/java/X3DJSAIL.html" target="_blank">X3D Java Scene Access Interface Library (X3DJSAIL)</a>.
		It has been produced using the 
		<a href="https://www.web3d.org/x3d/stylesheets/X3dToNodeJS.xslt" target="_blank">X3dToNodeJS.xslt</a>
		stylesheet to create NodeJS source code from an <code>.x3d</code> scene.
	</p>

 */

function JoeExample
  /** Default constructor to create this object. */
  ()
  {
    /** The initialized model object, created within initialize() method. */
    this.x3dModel = {};
    this.initialize();
    return this;
  }
JoeExample.prototype = {
  /** Create and initialize the X3D model. */
  initialize : function ()
  {
  this.x3dModel = (new autoclass.X3D()).setProfile("Immersive").setVersion("3.0")
  .setHead((new autoclass.head())
    .addMeta((new autoclass.meta()).setName("filename").setContent("Scripting.x3d")))
  .setScene((new autoclass.Scene())
    .addChild((new autoclass.Script("OpenVault")).setSourceCode("\n" + 
"\n" + 
"      " + "\n" + 
"\n" + 
"      ecmascript:" + "\n" + 
"\n" + 
"      function combinationEntered (value) {" + "\n" + 
"\n" + 
"        unlocked = value;" + "\n" + 
"\n" + 
"      }" + "\n" + 
"\n" + 
"      function openVault(value) {" + "\n" + 
"\n" + 
"      if (unlocked) vaultUnlocked = value;" + "\n" + 
"\n" + 
"      }" + "\n")
      .addField((new autoclass.field()).setName("openVault").setType("SFTime").setAccessType("inputOnly"))
      .addField((new autoclass.field()).setName("combinationEntered").setType("SFBool").setAccessType("inputOnly"))
      .addField((new autoclass.field()).setName("vaultUnlocked").setType("SFTime").setAccessType("outputOnly"))
      .addField((new autoclass.field()).setName("unlocked").setType("SFBool").setAccessType("initializeOnly").setValue("false")))
    .addChild((new autoclass.Shape())
      .setAppearance((new autoclass.Appearance())
        .setMaterial((new autoclass.Material()).setDiffuseColor(1.0,0.0,0.0)))
      .setGeometry((new autoclass.Sphere())))
    .addChild((new autoclass.Sound()).setMaxBack(1000).setMaxFront(1000).setMinBack(1000).setMinFront(1000)
      .setSource((new autoclass.AudioClip("Click")).setDescription("clicking sound").setStopTime(1).setUrl(new autoclass.MFString("\"../resources/chandubabamusic1.wav\""))))
    .addChild((new autoclass.TouchSensor("TS")))
    .addChild((new autoclass.ROUTE()).setFromNode("TS").setFromField("isOver").setToNode("OpenVault").setToField("combinationEntered"))
    .addChild((new autoclass.ROUTE()).setFromNode("TS").setFromField("touchTime").setToNode("OpenVault").setToField("openVault"))
    .addChild((new autoclass.ROUTE()).setFromNode("OpenVault").setFromField("vaultUnlocked").setToNode("Click").setToField("startTime")));
  },
  // end of initialize() method


  /** Provide a shallow copy of the X3D model.
   * @return JoeExample model
   */
  getX3dModel : function()
  {	  
	  return this.x3dModel;
  },
  
  /** Indicate X3DJSAIL validation results for this X3D model.
   * @return validation results plus exception information, if any
   */
  validateSelf : function()
  {
	var validationResult = "";
	var  exceptionResult = "";
	try
	{
		this.initialize();
		
		if ((this.getX3dModel() == null) || (this.getX3dModel().getHead() == null))
		{
			validationResult = "empty scene, nothing to validate. " + this.x3dModel.validate();
			return validationResult;
		}
		validationResult += this.x3dModel.validate(); // walk entire tree to validate correctness
	}
	catch (e)
	{
		exceptionResult = e; // report exception failures, if any
	}
	if  (exceptionResult === "" && validationResult === "")
	     return "success";
	else
	{
		var returnMessage = "";
		if  (exceptionResult !== "" && validationResult !== "")
			returnMessage += "\n*** ";
		returnMessage += exceptionResult;
		if  (exceptionResult === "" && validationResult !== "")
			returnMessage = "\n" + returnMessage; // skip line before meta tags, etc.
		returnMessage += validationResult;
		return returnMessage;
	}
  },
    /** Default main() method provided for test purposes.
     * @param argv input parameters
     */
    main : function (argv)
    {
		var testObject = new JoeExample();
		console.log ("JoeExample execution self-validation test results: " + testObject.validateSelf());
	}
}
new JoeExample().main();

File: src/main/python/net/x3djsonld/data/JoeExample.py
###############################################
#
# Now available: developmental python x3d.py package on PyPi for import.
#   This approach greatly simplifies Python X3D deployment and use.
#   https://pypi.org/project/x3d
#
# Installation:
#       pip install x3d
# or
#       python -m pip install x3d
#
# Developer options for loading x3d package:
#
#    from x3d import *  # preferred approach, terser source that avoids x3d.* class prefixes
#
# or
#    import x3d         # traditional way to subclass x3d package, all classes require x3d.* prefix
#                       # but python source is very verbose, for example x3d.Material x3d.Shape etc.
#                       # X3dToPython.xslt stylesheet insertPackagePrefix=true supports this option.

from x3d import *

###############################################

newModel=X3D(version='3.0',profile='Immersive',
  head=head(
    children=[
    meta(name='filename',content='Scripting.x3d')]),
  Scene=Scene(
    children=[
    Script(DEF='OpenVault',
      field=[
      field(name='openVault',type='SFTime',accessType='inputOnly'),
      field(name='combinationEntered',type='SFBool',accessType='inputOnly'),
      field(name='vaultUnlocked',type='SFTime',accessType='outputOnly'),
      field(name='unlocked',type='SFBool',value=False,accessType='initializeOnly')],

      sourceCode="""
ecmascript:

      function combinationEntered (value) {

        unlocked = value;

      }

      function openVault(value) {

      if (unlocked) vaultUnlocked = value;

      }
"""),
    Shape(
      appearance=Appearance(
        material=Material(diffuseColor=(1,0,0))),
      geometry=Sphere(),),
    Sound(maxBack=1000,maxFront=1000,minBack=1000,minFront=1000,
      source=AudioClip(DEF='Click',stopTime=1,description='clicking sound',url=["../resources/chandubabamusic1.wav"])),
    TouchSensor(DEF='TS'),
    ROUTE(fromNode='TS',fromField='isOver',toNode='OpenVault',toField='combinationEntered'),
    ROUTE(fromNode='TS',fromField='touchTime',toNode='OpenVault',toField='openVault'),
    ROUTE(fromNode='OpenVault',fromField='vaultUnlocked',toNode='Click',toField='startTime')])
) # X3D model complete

###############################################
# Self-test diagnostics
###############################################

if        metaDiagnostics(newModel): # built-in utility method in X3D class
    print(metaDiagnostics(newModel))
# print('check newModel.XML() serialization...')
newModelXML= newModel.XML() # test export method XML() for exceptions during export
newModel.XMLvalidate()

try:
#   print('check newModel.VRML() serialization...')
    newModelVRML=newModel.VRML() # test export method VRML() for exceptions during export
    # print(prependLineNumbers(newModelVRML)) # debug
    print("Python-to-VRML export of VRML output successful")
except BaseException as err:
    print("*** Python-to-VRML export of VRML output failed:", err)
    if newModelVRML: # may have failed to generate
        print(prependLineNumbers(newModelVRML, err.lineno))

try:
#   print('check newModel.JSON() serialization...')
    newModelJSON=newModel.JSON() # test export method JSON() for exceptions during export
#   print(prependLineNumbers(newModelJSON)) # debug
    print("Python-to-JSON export of JSON output successful (still testing)")
except SyntaxError as err:
    print("*** Python-to-JSON export of JSON output failed:", err)
    if newModelJSON: # may have failed to generate
        print(prependLineNumbers(newModelJSON,err.lineno))

print("python x3d.py load and self-test complete")
